File: backend/app/schemas/recipe.py
from pydantic import BaseModel, ConfigDict, Field


class RecipeNutritionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    calories_kcal: float
    protein_g: float
    carbohydrate_g: float
    fat_g: float
    sodium_mg: float
    sugar_g: float


class RecipeListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    title: str
    description: str
    cuisine: str
    meal_type: str
    servings: int
    total_time_minutes: int
    dietary_tags: list[str]
    nutrition: RecipeNutritionResponse
    course: str | None = None


class RecipeCollectionResponse(BaseModel):
    items: list[RecipeListItemResponse]
    next_cursor: int | None = None


class RecipeIngredientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    normalized_name: str
    quantity: float | None
    unit: str | None
    preparation: str | None
    allergens: list[str]


class RecipeStepResponse(BaseModel):
    step_number: int = Field(ge=1)
    instruction: str


class RecipeDetailResponse(RecipeListItemResponse):
    ingredients: list[RecipeIngredientResponse]
    steps: list[RecipeStepResponse]
    # Release recipes carry meal types (and a course); curated ones leave both empty.
    meal_types: list[str] | None = None
    prep_time_minutes: int | None = None
    cook_time_minutes: int | None = None

File: frontend/app/types/recipe.ts
export interface RecipeNutrition {
  calories_kcal: number;
  protein_g: number;
  carbohydrate_g: number;
  fat_g: number;
  sodium_mg: number;
  sugar_g: number;
}

export interface RecipeListItem {
  id: number;
  slug: string;
  title: string;
  description: string;
  cuisine: string;
  meal_type: string;
  servings: number;
  total_time_minutes: number;
  dietary_tags: string[];
  nutrition: RecipeNutrition;
  course?: string | null;
}

export interface RecipeCollection {
  items: RecipeListItem[];
  next_cursor: number | null;
}

export interface RecipeIngredient {
  name: string;
  normalized_name: string;
  quantity: number | null;
  unit: string | null;
  preparation: string | null;
  allergens: string[];
}

export interface RecipeStep {
  step_number: number;
  instruction: string;
}

export interface RecipeDetail extends RecipeListItem {
  ingredients: RecipeIngredient[];
  steps: RecipeStep[];
}

export interface TutorialVideo {
  video_id: string;
  title: string;
  channel_title: string;
  watch_url: string;
  embed_url: string;
  thumbnail_url: string | null;
  duration_seconds: number | null;
}

export interface TutorialRecommendation {
  recipe_slug: string;
  recipe_title: string;
  selected_video: TutorialVideo | null;
  retrieval: { provider_used: "youtube" | "fixture"; mode: string; status: string };
  warning: string | null;
}
